tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,342 @@
+		return false
+	}
+	return true
+}
+
+func checkForFailingTraffic(teaURL, coffeeURL string) bool {
+	if err := expectRequestToFail(teaURL, address, "URI: /tea"); err != nil {
+		return false
+	}
+	if err := expectRequestToFail(coffeeURL, address, "URI: /coffee"); err != nil {
+		return false
+	}
+	return true
+}
+
+func expectRequestToSucceed(appURL, address string, responseBodyMessage string) error {
+	status, body, err := framework.Get(appURL, address, timeoutConfig.RequestTimeout)
+	if status != http.StatusOK {
+		return errors.New("http status was not 200")
+	}
+
+	if !strings.Contains(body, responseBodyMessage) {
+		return errors.New("expected response body to contain correct body message")
+	}
+
+	return err
+}
+
+func expectRequestToFail(appURL, address string, responseBodyMessage string) error {
+	status, body, err := framework.Get(appURL, address, timeoutConfig.RequestTimeout)
+	if status != 0 {
+		return errors.New("expected http status to be 0")
+	}
+
+	if body != "" {
+		return fmt.Errorf("expected response body to be empty, instead received: %s", body)
+	}
+
+	if err == nil {
+		return errors.New("expected request to error")
+	}
+
+	return nil
+}
+
+// checkContainerLogsForErrors checks both nginx and ngf container's logs for any possible errors.
+// Since this function retrieves all the logs from both containers and the NGF pod may be shared between tests,
+// the logs retrieved may contain log messages from previous tests, thus any errors in the logs from previous tests
+// may cause an interference with this test and cause this test to fail.
+func checkContainerLogsForErrors(ngfPodName string) {
+	logs, err := resourceManager.GetPodLogs(
+		ngfNamespace,
+		ngfPodName,
+		&core.PodLogOptions{Container: nginxContainerName},
+	)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(logs).ToNot(ContainSubstring("[error]"), logs)
+	Expect(logs).ToNot(ContainSubstring("[crit]"), logs)
+	Expect(logs).ToNot(ContainSubstring("[alert]"), logs)
+	Expect(logs).ToNot(ContainSubstring("[emerg]"), logs)
+
+	logs, err = resourceManager.GetPodLogs(
+		ngfNamespace,
+		ngfPodName,
+		&core.PodLogOptions{Container: ngfContainerName},
+	)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(logs).ToNot(ContainSubstring("\"level\":\"error\""), logs)
+}
+
+func checkLeaderLeaseChange(originalLeaseName string) bool {
+	leaseName, err := getLeaderElectionLeaseHolderName()
+
+	if err != nil {
+		return false
+	}
+
+	if originalLeaseName != leaseName {
+		return true
+	}
+
+	return false
+}
+
+func getLeaderElectionLeaseHolderName() (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.GetTimeout)
+	defer cancel()
+
+	var lease coordination.Lease
+	key := types.NamespacedName{Name: "ngf-test-nginx-gateway-fabric-leader-election", Namespace: ngfNamespace}
+
+	if err := k8sClient.Get(ctx, key, &lease); err != nil {
+		return "", errors.New("could not retrieve leader election lease")
+	}
+	return *lease.Spec.HolderIdentity, nil
```

Comment:
what if `lease.Spec.HolderIdentity` is nil?

Reply:
I just checked and it never seems to be "". Though I will add a check for it.